numba/np/arraymath.py

```diff
@@ -4159,6 +4159,27 @@ def np_select_arr_impl(condlist, choicelist, default=0):
 
     return np_select_arr_impl
 
+
+@overload(np.asarray_chkfinite)
+def np_asarray_chkfinite(a, dtype=None, order='C'):
+
+    msg = "The argument to np.asarray_chkfinite must be array-like"
+    if not isinstance(a, (types.Array, types.Sequence, types.Tuple)):
+        raise TypingError(msg)
+
+    if is_nonelike(dtype):
+        dt = a.dtype
+    else:
+        dt = dtype.dtype
+
+    def impl(a, dtype=None, order='C'):
```

Comment:
I would suggest a different approach for the implementation. For the given, array-oriented approach of `np.all(np.isfinite(a))` I would assume we may loop through the array twice. Once to check each element for infinity and then to check if the result is True or False, which may involve an additional array to hold the booleans. So, perhaps the following is a better approach (sketch):

```
for i in a:
    if np.isfinite(a) or np.isnan(a):
        msg = "..."
        raise ValueError(msg)
# none of the values were NaN or infinity, we can safely create the array
return np.asarray(a, dtype=dt)

```

This uses a loop, but in contrast to regular Python, in Numba loops are a good practice and usually very fast. Also, you can then check both conditions while visiting each element only once.
Additionally also have the advantage of not creating the array unless it doesn't have a NaN or infinity.

Reply:
it may be worth timing the two approaches also, to ensure that the loop based one is indeed quicker.

Reply:
I tried this one but this is giving me an error that I am not able to understand.
```python
def impl(a, dtype=None, order='C'):
        for i in np.nditer(a):
            if np.isnan(i) or np.isinf(i):
                raise ValueError("array must not contain infs or NaNs")
        return np.asarray(a, dtype=dt)
```
error:
```
ERROR: test_asarray_chkfinite (numba.tests.test_np_functions.TestNPFunctions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "numba\numba\tests\test_np_functions.py", line 3757, in test_asarray_chkfinite
    got = cfunc(*pair)
  File "numba\numba\core\dispatcher.py", line 414, in _compile_for_args
    error_rewrite(e, 'typing')
  File "numba\numba\core\dispatcher.py", line 357, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<function asarray_chkfinite at 0x00000234879B4488>) found for signature:

 >>> asarray_chkfinite(reflected list(int64)<iv=None>, none, Literal[str](C))

There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'np_asarray_chkfinite': File: numba\np\arraymath.py: Line 4163.
    With argument(s): '(reflected list(int64)<iv=None>, none, unicode_type)':
   Rejected as the implementation raised a specific error:
     LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
   Invalid store of {i8*, i8*}* to i64* in <numba.core.datamodel.models.ArrayModel object at 0x000002348B55C128> (trying to write member #4)

   File "numba\np\arraymath.py", line 4176:
       def impl(a, dtype=None, order='C'):
           for i in np.nditer(a):
           ^

   During: lowering "$14.2 = iternext(value=$phi14.1)" at numba\numba\np\arraymath.py (4176)
  raised from numba\numba\core\errors.py:745

During: resolving callee type: Function(<function asarray_chkfinite at 0x00000234879B4488>)
During: typing of call at numba\numba\tests\test_np_functions.py (325)


File "numba\tests\test_np_functions.py", line 325:
def np_asarray_chkfinite(a, dtype=None, order='C'):
    return np.asarray_chkfinite(a, dtype, order)
    ^


----------------------------------------------------------------------
Ran 1 test in 1.064s

FAILED (errors=1)
```

Reply:
Interesting, my guess would be to try `for i in a:` instead of `for i in np.nditer(a)`.

Reply:
```for i in a``` will work for 1D array but for n-dim array ```if np.isinf(i) or np.isnan(i)``` will give an error as np.isinf and np.isnan returns True/False for each value (boolean array) and we have to use .any() to check if there is any True value. Maybe I can use flatten here? or did I I misunderstood?

Reply:
That is indeed a tricky situation to be in. Since the valid input to this function can be either an array or a sequence, like a list or even a tuple (which may even be heterogeneously typed?). It seems like the implementation for this might be more complex than initially anticipated. There are now a few routes here and it isn't clear to me, which is the best:

a) Return to the original implementation of doing `asarray` first and then checking for inf or NaN. This would make the code shorter and the implementation easy, but may incur unneeded conversion to an array. The advantage is that Numpy will take care of converting the potentially deeply nested data-structures for you. You can still opt to iterate over the final array and bail on the first sign of inf or NaN though.

For example like so:

```
def impl(a, dtype=None, order='C'):
        a = np.asarray(a, dtype=dt)
        for i in np.nditer(a):
            if np.isnan(i) or np.isinf(i):
                raise ValueError("array must not contain infs or NaNs")
        return a
```

b) Specialize on different types of input. I.e. have different implementations for arrays/tuples/lists. However, this might end up being quite contrived as the types may be nested within each other:

```
In [7]: l = List((tuple((np.zeros(2),np.ones(2),)), tuple((np.zeros(2),np.ones(2),))))

In [8]: l
Out[8]: ListType[UniTuple(array(float64, 1d, C) x 2)]([(array([0., 0.]), array([1., 1.])), (array([0., 0.]), array([1., 1.]))])

In [9]: np.asarray(l)
Out[9]:
array([[[0., 0.],
        [1., 1.]],

       [[0., 0.],
        [1., 1.]]])
```

In any case, after thinking about this some more doing option a) may actually be the better choice here. Apologies for not seeing that before.

It may also be worth looking at what the Numpy implementation does and then attempt to replicate that in case you haven't already done so.

Reply:
I have checked Numpy [implementation](https://github.com/numpy/numpy/blob/v1.19.0/numpy/lib/function_base.py#L422-L487) they used ```np.isfinite().all()``` instead of using the for loop and isnan or isinf.

Reply:
@rishabhvarshney14 great find. Then do that!

Reply:
Actually my first implementation is based on the numpy's implementation.
```
def impl(a, dtype, order):
    if not np.all(np.isfinite(a)):
        raise_error
    return a
```
which is slower than using for loop with np.isnan and np.isinf
```
def impl(a, dtype=None, order='C'):
    a = np.asarray(a, dtype=dt)
    for i in np.nditer(a):
        if np.isnan(i) or np.isinf(a):
            raise ValueError("array must not contain infs or NaNs")
    return a
```

Reply:
Wow, even better! You could try to see if `if np.isnan(i) or np.isinf(a)` is slower/faster than np.isfinite(a). Also, what sort of timings are you seeing? Is it 10%, 20% faster? Or even more? I think it also depends largely on where in the array the `Nan` values are. The second variant you have should be faster when the arrays have `Nan` at the beginning but about equal if there is only one `Nan` and it is the last element of the array.

Reply:
I think using for loop is around 30-40 % faster than np.isfinite but I might be wrong also isfinite is much slower when there is a nan in beginning since it checks the whole array. Even when array has nan value in last np.isfinite is slower.
You can check it [here](https://colab.research.google.com/drive/1qjdM6dT5BMv_13gqaWwGwaSRDfhIotTr?usp=sharing).

Reply:
You can also use `np.isfinite` within the for-loop on a single element.

Reply:
I've conducted some benchmarks myself to gather more evidence, I'll post them in the main thread of this PR.